p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,756 @@
+
+		}
+		return out.host == "" || out.port == "" || !out.useHTTPS || out.sni == ""
+	})
+
+	if out.useHTTPS && out.sni == "" {
+		out.sni = out.host
+	}
+	return out
+}
+
+var httpComponent, _ = ma.NewComponent("http", "")
+var tlsComponent, _ = ma.NewComponent("tls", "")
+
+// normalizeHTTPMultiaddr converts an https multiaddr to a tls/http one.
+// Returns a bool indicating if the input multiaddr has an http (or https) component.
+func normalizeHTTPMultiaddr(addr ma.Multiaddr) (ma.Multiaddr, bool) {
+	isHTTPMultiaddr := false
+	beforeHTTPS, afterIncludingHTTPS := ma.SplitFunc(addr, func(c ma.Component) bool {
+		if c.Protocol().Code == ma.P_HTTP {
+			isHTTPMultiaddr = true
+		}
+
+		if c.Protocol().Code == ma.P_HTTPS {
+			isHTTPMultiaddr = true
+			return true
+		}
+		return false
+	})
+
+	if afterIncludingHTTPS == nil {
+		// No HTTPS component, just return the original
+		return addr, isHTTPMultiaddr
+	}
+
+	_, afterHTTPS := ma.SplitFirst(afterIncludingHTTPS)
+	if afterHTTPS == nil {
+		return ma.Join(beforeHTTPS, tlsComponent, httpComponent), isHTTPMultiaddr
+	}
+
+	return ma.Join(beforeHTTPS, tlsComponent, httpComponent, afterHTTPS), isHTTPMultiaddr
+}
+
+// ProtocolPathPrefix looks up the protocol path in the well-known mapping and
+// returns it. Will only store the peer's protocol mapping if the server ID is
+// provided.
+func (h *Host) getAndStorePeerMetadata(roundtripper http.RoundTripper, server peer.ID) (PeerMeta, error) {
+	if h.peerMetadata == nil {
+		h.peerMetadata = newPeerMetadataCache()
+	}
+	if meta, ok := h.peerMetadata.Get(server); server != "" && ok {
+		return meta, nil
+	}
+
+	req, err := http.NewRequest("GET", "/.well-known/libp2p", nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Accept", "application/json")
+
+	client := http.Client{Transport: roundtripper}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	body := [peerMetadataLimit]byte{}
+	bytesRead := 0
+	for {
+		n, err := resp.Body.Read(body[bytesRead:])
+		bytesRead += n
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		if bytesRead >= peerMetadataLimit {
+			return nil, fmt.Errorf("peer metadata too large")
+		}
+	}
+
+	meta := PeerMeta{}
+	json.Unmarshal(body[:bytesRead], &meta)
+	if server != "" {
+		h.peerMetadata.Add(server, meta)
+	}
+
+	return meta, nil
+}
+
+// AddPeerMetadata adds a peer's protocol metadata to the http host. Useful if
+// you have out-of-band knowledge of a peer's protocol mapping.
+func (h *Host) AddPeerMetadata(server peer.ID, meta PeerMeta) {
+	if h.peerMetadata == nil {
+		h.peerMetadata = newPeerMetadataCache()
+	}
+	h.peerMetadata.Add(server, meta)
+}
+
+// GetPeerMetadata gets a peer's cached protocol metadata from the http host.
+func (h *Host) GetPeerMetadata(server peer.ID) (PeerMeta, bool) {
```

Comment:
Is there any reason to return the whole map here? Alternative: `SupportsProtocol(peer.ID) (path string, bool)`. It would be nice if we didn't have to make the map an exported type.

Reply:
The reason to export the whole type is to give users flexibility in managing the Host's peerMetadata cache without exposing that cache. This is useful if they want to add out of band knowledge of a peer's metadata. This method along with Set and Remove should let users do anything they need to with this metadata cache.